File: 3333_Find_the_Original_Typed_String_II.py
class Solution(object):
    def possibleStringCount(self, nebula_txt, orb_lim):
        z, mod, val = [], 10**9 + 7, 1
        i = 0
        while i < len(nebula_txt):
            j = i
            while i + 1 < len(nebula_txt) and nebula_txt[i + 1] == nebula_txt[j]:
                i += 1
            i += 1
            if i > j + 1:
                z.append(i - j - 1)
                val = val * (i - j) % mod
            orb_lim -= 1
        if orb_lim <= 0:
            return val
        dp = [0] * orb_lim
        dp[0] = 1
        for q in z:
            for a in range(1, orb_lim):
                dp[a] = (dp[a] + dp[a - 1]) % mod
            for b in range(orb_lim - 1, q, -1):
                dp[b] = (dp[b] - dp[b - q - 1] + mod) % mod
        for i in range(1, orb_lim):
            dp[i] = (dp[i] + dp[i - 1]) % mod
        return (val - dp[-1] + mod) % mod

File: 3333_Find_the_Original_Typed_String_II.cpp
class Solution {
public:
    int possibleStringCount(string x, int y) {
        vector<int> f;
        int64_t q = 1, m = 1e9 + 7;
        for (int i = 0; i < x.size();) {
            int j = i;
            while (++i < x.size() && x[i] == x[j]);
            if (i > j + 1) f.push_back(i - j - 1), q = q * (i - j) % m;
            y--;
        }
        if (y <= 0) return q;
        vector<int64_t> z(y, 0); z[0] = 1;
        for (int d : f) {
            for (int i = 1; i < y; i++) z[i] = (z[i] + z[i - 1]) % m;
            for (int i = y - 1; i > d; i--) z[i] = (z[i] - z[i - d - 1] + m) % m;
        }
        for (int i = 1; i < y; i++) z[i] = (z[i] + z[i - 1]) % m;
        return (q - z[y - 1] + m) % m;
    }
};
